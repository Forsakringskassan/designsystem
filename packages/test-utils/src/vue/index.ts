export * from "../lib";
export {
    SizeWrapper,
    sizeWrapperHeight,
    sizeWrapperWidth,
} from "./size-wrapper";
export {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "./density-wrapper";
