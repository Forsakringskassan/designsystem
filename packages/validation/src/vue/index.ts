export { type ValidationDirective } from "./validation-directive";
export { ValidationPlugin } from "./validation-plugin";
export {
    type UseValidation,
    type UseValidationOptions,
    type ValidityModel,
    useValidation,
} from "./use-validation";
export { useValidationConfig } from "./use-validation-config";
