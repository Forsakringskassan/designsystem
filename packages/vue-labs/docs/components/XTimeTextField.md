---
title: Tidsinmatning
status: Experimentell
layout: component
component: XTimeTextField
---

## Användning

När det behövs ett textinmatningsfält för timmar och minuter, tt:mm. Accepterar inmatning med eller
utan kolon. Utan kolon tolkas inmatningen som enbart timmar. Om man matar in ett kolon måste det
följas av två siffror för att vara giltigt (dvs. 08: och 08:0 är ogiltiga). Exempel:

| Inmatning | Formattering | Värde |
| --------- | ------------ | ----- |
| 1234      | 1234:00      | 74040 |
| 123       | 123:00       | 7380  |
| 12        | 12:00        | 720   |
| 8         | 08:00        | 480   |
| 12:30     | 12:30        | 750   |

Lagrar värdet internt i ett `Number` som plockas ut med `v-model`.

### Enkelt exempel

```import
XTimeTextFieldExample.vue
```

## Förlåtande inmatning

Det är inte ovanligt med ett behov av att endast tillåta tidsangivelser upp till ett dygn
(24 timmar). Det går att sätta en egen parsingparameter istället. Detta görs i stil med
`:parser="forgivingParseTimeToNumber"`. I denna komponenten tolkas även fyra siffror inte
längre som timmar, utan som timmar och minuter. De två första siffrorna tolkas som timmar
och de två sista som minuter. Notera att tre siffror och fem siffror fortfarande tolkas
som timmar.
Exempel:

| Inmatning | Formattering | Värde |
| --------- | ------------ | ----- |
| 1234      | 12:34        | 74040 |
| 123       | 123:00       | 7380  |
| 12        | 12:00        | 720   |
| 8         | 08:00        | 480   |
| 12:30     | 12:30        | 750   |

### Exempel med förlåtande inmatning

```import
ForgivingInput.vue
```
