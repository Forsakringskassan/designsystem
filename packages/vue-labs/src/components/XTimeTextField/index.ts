export * from "./utils";
export * from "./converters";
export type * from "./validators";
export { default as XTimeTextField } from "./XTimeTextField.vue";
