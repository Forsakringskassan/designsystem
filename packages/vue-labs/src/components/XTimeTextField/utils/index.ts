export * from "./constants";
export * from "./HoursMinutesNumberUtils";
export * from "./HoursMinutesValidatorUtils";
