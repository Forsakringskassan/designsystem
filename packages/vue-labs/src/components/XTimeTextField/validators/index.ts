import { type Validator, ValidationService } from "@fkui/logic";
import { greaterThanTimeValidator } from "./GreaterThanTimeValidator";
import { hoursMinutesValidator } from "./HoursMinutesValidator";
import { lessThanTimeValidator } from "./LessThanTimeValidator";
import { maxTimeValidator } from "./MaxTimeValidator";
import { minTimeValidator } from "./MinTimeValidator";

const validators: Validator[] = [
    hoursMinutesValidator,
    greaterThanTimeValidator,
    lessThanTimeValidator,
    maxTimeValidator,
    minTimeValidator,
];

for (const validator of validators) {
    ValidationService.registerValidator(validator);
}
