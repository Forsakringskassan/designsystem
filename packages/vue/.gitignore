# vite dev
/src/vite-dev/app.vue
