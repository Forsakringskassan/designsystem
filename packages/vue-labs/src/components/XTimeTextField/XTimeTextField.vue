<script lang="ts">
import { type PropType, defineComponent } from "vue";
import { FTextField, TranslationMixin, type ParseFunction, type FormatFunction, useTextFieldSetup } from "@fkui/vue";
import { isSet, ValidationService } from "@fkui/logic";
import { HoursMinutesString, formatNumberToTime, parseTimeToNumber } from "./converters";
import "./validators";

export default defineComponent({
    name: "XTimeTextField",
    extends: FTextField,
    mixins: [TranslationMixin],
    props: {
        formatter: {
            type: Function as PropType<FormatFunction<HoursMinutesString>>,
            required: false,
            default: formatNumberToTime,
        },
        parser: {
            type: Function as PropType<ParseFunction<number>>,
            required: false,
            default: parseTimeToNumber,
        },
    },
    setup(props) {
        return useTextFieldSetup(props);
    },
    mounted() {
        const inputElement: HTMLInputElement = this.$el.querySelector("input");

        if (!isSet(inputElement)) {
            throw new Error(`Could not find input element in XTimeTextField with id ${this.$el.id}`);
        }

        ValidationService.addValidatorsToElement(
            inputElement,
            {
                maxLength: { length: 10 },
                hoursMinutes: {},
            },
            true,
        );

        inputElement.setAttribute("inputmode", "numeric");

        ValidationService.validateElement(inputElement);
    },
});
</script>
