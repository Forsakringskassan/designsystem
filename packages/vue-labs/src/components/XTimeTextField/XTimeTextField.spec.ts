import { defineComponent } from "vue";
import { VueWrapper, mount } from "@vue/test-utils";
import { ValidationPlugin } from "@fkui/vue";
import XTimeTextField from "./XTimeTextField.vue";

function createWrapper(
    template: string,
    myModel: string | number,
): VueWrapper<InstanceType<typeof TestComponent>> {
    const TestComponent = defineComponent({
        name: "TestComponent",
        components: {
            XTimeTextField,
        },
        data() {
            return {
                myModel,
            };
        },
        template,
    });

    return mount(TestComponent, {
        global: {
            plugins: [ValidationPlugin],
        },
    });
}

describe("snapshots", () => {
    it("should match snapshot with label and input", () => {
        const wrapper = createWrapper(
            /* HTML */ `
                <x-time-text-field id="myField" v-model="myModel">
                    My hours minutes field
                </x-time-text-field>
            `,
            "",
        );
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("v-model", () => {
    it("should update model with correct parse number value from time value input", async () => {
        const wrapper = createWrapper(
            /* HTML */ `
                <x-time-text-field id="myField" v-model="myModel">
                    My hours minutes field
                </x-time-text-field>
            `,
            "",
        );

        const input = wrapper.get("input");
        input.setValue("12:00");
        await input.trigger("blur");

        expect(wrapper.vm.$data.myModel).toBe(720);
    });
});

describe("format", () => {
    it("should format to correct time value", async () => {
        const wrapper = createWrapper(
            /* HTML */ `
                <x-time-text-field id="myField" v-model="myModel">
                    My hours minutes field
                </x-time-text-field>
            `,
            "",
        );

        const input = wrapper.get("input");
        const inputElement = input.element as HTMLInputElement;

        input.setValue("1 2:35");
        await input.trigger("blur");
        await wrapper.vm.$nextTick();

        expect(inputElement.value).toBe("12:35");
    });
});
