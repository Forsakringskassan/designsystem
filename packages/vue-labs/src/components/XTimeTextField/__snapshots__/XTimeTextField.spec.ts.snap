// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot with label and input 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="myField"
    >
      
      
       My hours minutes field 
      
      
       
      
      
      <!---->
       
      <!---->
      
      
       
      <!---->
    </label>
  </div>
   
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
     
    <div
      class="text-field__icon-wrapper"
    >
      <input
        aria-invalid="false"
        class="text-field__input"
        data-validation=""
        id="myField"
        inputmode="numeric"
        type="text"
      />
       
      <!---->
       
      <!---->
       
      <!---->
       
      <!---->
    </div>
     
    
    
  </div>
   
  <!---->
</div>
`;
